entity-service-client/src/main/java/org/hypertrace/entity/query/service/client/EntityLabelsCachingClient.java

```diff
@@ -0,0 +1,216 @@
+    try {
+      Map<EntityCacheKey<EntityTypeAndId>, List<String>> entityCacheKeyToLabelsMap =  entityIdsToLabelsCache.getAll(entityCacheKeys);
+      return entityCacheKeyToLabelsMap.entrySet().stream()
+          .collect(
+              Collectors.toMap(entry -> entry.getKey().getDataKey().getId(), Entry::getValue)
+          );
+    } catch (ExecutionException ex) {
+      throw new RuntimeException("EntityLabelsCachingClient exception while getting labels for entities", ex);
+    }
+  }
+
+  private Map<EntityCacheKey<EntityTypeAndId>, List<String>> loadEntityLabelsForKeys(Iterable<? extends EntityCacheKey<EntityTypeAndId>> keys) {
+    Iterator<? extends EntityCacheKey<EntityTypeAndId>> iter = keys.iterator();
+    if (!iter.hasNext()) {
+      return Map.of();
+    }
+
+    List<EntityCacheKey<EntityTypeAndId>> entityIdKeys = new ArrayList<>();
+    while (iter.hasNext()) {
+      entityIdKeys.add(iter.next());
+    }
+
+    // The assumption we make is that the all the entity id keys are of the same entity type and
+    // belong to the same tenant. So we can extract those values from the first cache key.
+    EntityCacheKey<EntityTypeAndId> firstEntityCacheKey = entityIdKeys.get(0);
+    String entityType = firstEntityCacheKey.getDataKey().getType();
+    String entityIdColumnName = firstEntityCacheKey.getDataKey().getIdColumnName();
+    String entityLabelsColumnName = firstEntityCacheKey.getDataKey().getLabelsColumnName();
+    Map<String, String> headers = firstEntityCacheKey.getHeaders();
+
+    Map<String, EntityCacheKey<EntityTypeAndId>> entityIdToCacheKey = new HashMap<>();
+    List<String> entityIds = entityIdKeys.stream()
+        .map(key -> {
+          entityIdToCacheKey.put(key.getDataKey().getId(), key);
+          return key.getDataKey().getId();
+        })
+        .collect(Collectors.toList());
+
+    EntityQueryRequest entityQueryRequest = EntityQueryRequest.newBuilder()
+        .setEntityType(entityType)
+        .addSelection(createColumnExpression(entityIdColumnName))
+        .addSelection(createColumnExpression(entityLabelsColumnName))
+        .setFilter(
+            Filter.newBuilder()
+                .setOperator(Operator.AND)
+                .addChildFilter(
+                    Filter.newBuilder()
+                        .setOperator(Operator.IN)
+                        .setLhs(createColumnExpression(entityIdColumnName))
+                        .setRhs(createStringArrayLiteralExpression(entityIds)
+                        )
+                )
+        )
+        .setLimit(entityIds.size())
+        .build();
+
+    Iterator<ResultSetChunk> resultSetChunkIterator =
+        entityQueryServiceClient.execute(entityQueryRequest, headers);
+
+    Map<EntityCacheKey<EntityTypeAndId>, List<String>> entityLabelsByEntityIdMap = new HashMap<>();
+    while (resultSetChunkIterator.hasNext()) {
+      ResultSetChunk chunk = resultSetChunkIterator.next();
+
+      if (chunk.getRowCount() < 1) {
```

Comment:
Is this needed? It avoids calling the for loop on an empty list, but doesn't change the logic otherwise.

Reply:
Not necessarily needed but makes it more readable as you know you can stop reading here.